lib/action_view/component/render_monkey_patch.rb

```diff
@@ -15,15 +15,34 @@ def render(options = {}, args = {}, &block)
 
           options.render_in(self, &block)
         elsif options.is_a?(Class) && options < ActionView::Component::Base
-          options.new(args).render_in(self, &block)
+          safe_new_component(options, args).render_in(self, &block)
         elsif options.is_a?(Hash) && options.has_key?(:component)
-          options[:component].new(options[:locals]).render_in(self, &block)
+          safe_new_component(options[:component], options[:locals]).render_in(self, &block)
         elsif options.respond_to?(:to_component_class) && !options.to_component_class.nil?
+          # In this case, the `options` passed to `new` is the object on which `#to_component_class` is called. So we
+          # want the non-zero arity initializer to be called.
           options.to_component_class.new(options).render_in(self, &block)
         else
           super
         end
       end
+
+      private
+
+      def safe_new_component(component_class, options)
```

Comment:
As this behavior will never change based on the state of the component, why don't we compile this into a method on the component that has the correct arity handling?

Reply:
I'm probably wrong about this, but I'm not sure we know the arity of the initialize method when we inherit from `Base`. I think we have to check the arity when the component is initialized. 

What do you think of moving this logic in to a new method,`Base::safe_new`, that the monkey patch calls?

Reply:
I'm giving this a shot but sorta getting lost. Maybe we should set up some time to pair/chat synchronously about this. 

Here are some things that are tripping me up: 
* Since `compile!` is called after a component is rendered, it's too late to define the "safe initializer", as the component is already initialized.
* Since we currently inspect the `options` to avoid dropping them, there is _some_ behavior change based on the state of the component. I thought that maybe raising our own error in this situation would work.

Here are the changes to compile that I was considering, but wont work (because compile is called after the first time `initialize` is called).

```rb
        def compile(validate: false)
          return if compiled?

          if template_errors.present?
            raise ActionView::Component::TemplateError.new(template_errors) if validate
            return false
          end

          templates.each do |template|
            class_eval <<-RUBY, __FILE__, __LINE__ + 1
              def #{call_method_name(template[:variant])}
                @output_buffer = ActionView::OutputBuffer.new
                #{compiled_template(template[:path])}
              end
            RUBY
          end

          # To allow component initializers that take no options,
          # check the arity of `#initialize` to avoid passing an
          # empty `options` hash when no arguments are expected.
          if instance_method(:initialize).arity.zero?
            safe_new_method = <<-RUBY
              def safe_new(options)
                unless options.blank? # Check `options` to avoid dropping arguments passed to `#render`
                  raise "Component `#{name}` was rendered with options, but initializer does not accept arguments."
                end
  
                new
              end
            RUBY
          else
            safe_new_method = <<-RUBY
              def safe_new(options)
                new(options)
              end
            RUBY
          end

          class_eval safe_new_method, __FILE__, __LINE__ + 1

          @compiled = true
        end
```

Taking a step back, I was wondering what was behind this suggestion, were you thinking that it was best not to add additional logic to render/initialization if possible?